default.nix

```diff
@@ -1,22 +1,92 @@
-{ pkgs ? import <nixpkgs> { }
-, hypothesisProfile ? null
-, collectCoverage ? false
-, testSuite ? null
-, trialArgs ? null
-, tahoe-lafs ? "1.14.0"
-}:
 let
-  tahoe-packages = {
-    "1.14.0"    = pkgs.python2Packages.tahoe-lafs-1_14;
-    "1.16.0rc1" = pkgs.python2Packages.callPackage ./nix/tahoe-lafs-1_16.nix { };
-  };
-  tahoe-lafs' = builtins.getAttr tahoe-lafs tahoe-packages;
-
-  pkgs' = pkgs.extend (import ./overlays.nix);
-  callPackage = pkgs'.python27Packages.callPackage;
+  sources = import nix/sources.nix;
 in
-callPackage ./zkapauthorizer.nix {
-  challenge-bypass-ristretto = callPackage ./python-challenge-bypass-ristretto.nix { };
-  inherit hypothesisProfile collectCoverage testSuite trialArgs;
-  tahoe-lafs = tahoe-lafs';
-}
+{ pkgs ? import sources.release2015 {}
+, pypiData ? sources.pypi-deps-db
+, mach-nix ? import sources.mach-nix { inherit pkgs pypiData; }
+, tahoe-lafs-source ? "tahoe-lafs"
+, tahoe-lafs-repo ? sources.${tahoe-lafs-source}
+}:
+  let
+    lib = pkgs.lib;
+    python = "python27";
+    providers = {
+      _default = "sdist,nixpkgs,wheel";
+      # mach-nix doesn't provide a good way to depend on mach-nix packages,
+      # so we get it as a nixpkgs dependency from an overlay. See below for
+      # details.
+      tahoe-lafs = "nixpkgs";
+      # not packaged in nixpkgs at all, we can use the binary wheel from
+      # pypi though.
+      python-challenge-bypass-ristretto = "wheel";
+      # Pure python packages that don't build correctly from sdists
+      # - patches in nixpkgs that don't apply
+      boltons = "wheel";
+      chardet = "wheel";
+      urllib3 = "wheel";
+      # - incorrectly detected dependencies due to pbr
+      fixtures = "wheel";
+      testtools = "wheel";
+      traceback2 = "wheel";
+      # - Incorrectly merged extras - https://github.com/DavHau/mach-nix/pull/334
+      tqdm = "wheel";
+    };
+  in
+    rec {
+      tahoe-lafs = mach-nix.buildPythonPackage rec {
+        inherit python providers;
+        name = "tahoe-lafs";
+        version = "1.16.post999";
+        # See https://github.com/DavHau/mach-nix/issues/190
+        requirementsExtra = ''
+          pyrsistent < 0.17
+          foolscap == 0.13.1
+          configparser
+          eliot
+        '';
+        postPatch = ''
+          cat > src/allmydata/_version.py <<EOF
+          # This _version.py is generated by nix.
+
+          verstr = "${version}+git-${tahoe-lafs-repo.rev}"
+          __version__ = verstr
+          EOF
+        '';
+        src = tahoe-lafs-repo;
+      };
+      zkapauthorizer = mach-nix.buildPythonApplication rec {
+        inherit python providers;
+        src = ./.;
```

Comment:
I thought that `src = lib.cleanSource ./.;` would be a nice improvement here to improve cache hit rate.

I can't catch `nix build` actually /building/ anything.  It just spends 1m2s evaluating Nix expressions and then is done (if there's a cache hit) I guess?  That's a slightly disappointing evaluating time for the cached case. :(  (You mentioned how mach-nix is similar to haskell.nix and this would be another point of similarity - cached haskell.nix builds also take quite a while to evaluate).

It does help with the test suite, at least.  Though it still pays the 1m penalty for evaluating default.nix, it improves cache hit rate and so increases the chances of skipping the 4 minutes of actually running the tests.